### ISSUE TITLE
Crashes with RangeError during streaming to S3

### ISSUE THREAD

**Issue by tleyendecker** [ksoichiro/node-archiver-zip-encryptable#1]
I used archiver to realise a full streaming solution from a storage to a storage (S3).

S3 -> compress -> S3

After I added 'archiver-zip-encryptable' I run into a issue. I tested with small files (e.g 1 mb) everything was fin. With files larger than 20 mb the underlying zip-crypto-stream.js crashes with a RangeError.

 ```
The value of "value" is out of range. It must be >= 0 and <= 4294967295. Received 4294967479
RangeError [ERR_OUT_OF_RANGE]: The value of "value" is out of range. It must be >= 0 and <= 4294967295. Received 4294967479
    at checkInt (internal/buffer.js:36:11)
    at writeU_Int32BE (internal/buffer.js:628:3)
    at Buffer.writeUInt32BE (internal/buffer.js:641:10)
    at module.exports.ZipCrypto._updateKeys (/node_modules/archiver-zip-encryptable/lib/zip-crypto.js:106:16)
    at module.exports.ZipCrypto.encrypt (/node_modules/archiver-zip-encryptable/lib/zip-crypto.js:126:10)
    at module.exports.handleStuff (/node_modules/archiver-zip-encryptable/lib/zip-crypto-stream.js:79:24)
    at Object.onceWrapper (events.js:276:13)
    at module.exports.emit (events.js:193:15)
    at module.exports.EventEmitter.emit (domain.js:441:20)
    at endReadableNT (_stream_readable.js:1129:12)
    at process.internalTickCallback (internal/process/next_tick.js:72:19)
```
Is there any solution to fix that bug ?

**Comment by ksoichiro** [comment]
Thank you for your report!
I confirmed the issue.
I'll fix it soon.

**Comment by ksoichiro** [comment]
@tleyendecker I've just fixed the bug and published v1.0.1.
If you find the issue is not resolved, please reopen it.

**Comment by tleyendecker** [comment]
Thank you for the really fast response. 
Unfortunately, I'm getting another issue now. 

I tried to zip 20 files with 20 mb. Following module raised an error:
```
module.exports.handleStuff (/node_modules/archiver-zip-encryptable/lib/zip-crypto-stream.js:89:5)
```

**Comment by ksoichiro** [comment]
Thank you for your feedback.
Could you show me the error description?

**Comment by tleyendecker** [comment]
Hey, I'm not sure if it helps, after a closer look at the error I realised that the error raised from a "connection time out" from S3.
The interesting thing is, if I don't compress with a password, I'm able to zip Files > 1GB. I'm not sure, but I guess it is a bad combination of streaming delays.

```
{ TimeoutError: Connection timed out after 120000ms
  at ClientRequest.<anonymous> (/node_modules/aws-sdk/lib/http/node.js:83:34)
  at Object.onceWrapper (events.js:276:13)
  at ClientRequest.emit (events.js:188:13)
  at ClientRequest.EventEmitter.emit (domain.js:441:20)
  at Socket.emitRequestTimeout (_http_client.js:668:40)
  at Object.onceWrapper (events.js:276:13)
  at Socket.emit (events.js:188:13)
  at Socket.EventEmitter.emit (domain.js:441:20)
  at Socket._onTimeout (net.js:446:8)
  at listOnTimeout (timers.js:324:15)
  at processTimers (timers.js:268:5)
message: 'Connection timed out after 120000ms',
code: 'TimeoutError',
time: 2019-01-30T12:10:54.054Z,
region: 'eu-central-1',
hostname: '192.168.10.121',
retryable: false,
statusCode: 200,
retryDelay: 71.82894491080422 }
(node:4155) UnhandledPromiseRejectionWarning: Error: Unfortunately the service isn´t available at the moment
  at Archiver.archive.on (/src/lib/Archive.mjs:304:16)
  at Archiver.emit (events.js:193:15)
  at Archiver.EventEmitter.emit (domain.js:441:20)
  at Archiver.<anonymous> /node_modules/archiver/lib/core.js:187:12)
  at module.exports.handleStuff (/node_modules/archiver-zip-encryptable/lib/zip-crypto-stream.js:89:5)
  at Object.onceWrapper (events.js:276:13)
  at module.exports.emit (events.js:193:15)
  at module.exports.EventEmitter.emit (domain.js:441:20)
  at endReadableNT (_stream_readable.js:1129:12)
  at process.internalTickCallback (internal/process/next_tick.js:72:19)
```

**Comment by ksoichiro** [comment]
Thank you for your comment.

It may be caused by encrypting one file at a time.

Could you try again with the commit 7c22a2e5ad?
It's not published to npmjs. It can be installed with the following notation:

```json
  "dependencies": {
    "archiver-zip-encryptable": "git+https://github.com/ksoichiro/node-archiver-zip-encryptable.git#7c22a2e"
  }
```

**Comment by tleyendecker** [comment]
Hey me again, I have good and bad news for you. 
I run several test and you did it, the last bug is gone :).

Unfortunately there is a other issue.
I tried with 20x20mb files -> The compressed file was just 75mb large. After unziping 4 Files was ok, the rest of it broken.
A other test with 3x200mb files results ends up in a similar result. One file was ok, the rest broken.

Maybe you have a idea to fix it.

ps.: you do an amazing job =D

**Comment by ksoichiro** [comment]
Thank you so much for your testing!

> I tried with 20x20mb files -> The compressed file was just 75mb large. After unziping 4 Files was ok, the rest of it broken.

Hmm...that's weird. Unfortunately I still haven't been able to reproduce the issue, but I'll try on this weekend.

**Comment by ksoichiro** [comment]
@tleyendecker I made a sample script for AWS Lambda to read files from S3, compress, and upload zip file to S3. (See samples/s3 directory)

But unfortunately I couldn't reproduce it.

With this sample, I used text file that has the following content (24MB) and appended it 20 times to a zip file:

```
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
...
```

and password was `abcdef123456!`.
Zip file is successfully created with 3.2MB and can be unzipped without broken files.

What is different from your tests?

**Comment by tleyendecker** [comment]
Ok your test setup is more or less the same then mine. I used pdfs.

The oposite of your testing is the total file size (in MB). In my tests, I always used as many files that I end up in a minimum file size of 250 MB or more.
The zip was always succesfully created. But after I unziped, some files was broken. In the most cases the created zip was round about 75MB.

Hope it helps.

**Comment by ksoichiro** [comment]
Thanks. I converted text file to PDF and tried, but could not reproduce... :sob:
Do you set any options to archiver other than password?

```js
  archiver('zip-encryptable', {
    password: password
  });
```

**Comment by ksoichiro** [comment]
One more question. What version of Node.js are you using?

**Comment by tleyendecker** [comment]
hey, sorry for the late response. 

I tried your advice, and it works in most cases.  On really large file sizes (e.g. 1 GB) the AWS readStream always die with StreamContentLengthMismatch. But this is a other problem which I must deal with.

**Comment by ksoichiro** [comment]
Thanks for checking.
I'm glad that it worked for your case.

> But this is a other problem which I must deal with.

OK, I'll close this issue at this moment.
If you think that this package has a problem, feel free to reopen this or create a new issue.

BTW the fix that I mentioned above (7c22a2e) is already merged to master and published as v1.0.2.
